packages/stream/index.js

```diff
@@ -115,6 +115,91 @@ export const makePipe = () => {
 };
 harden(makePipe);
 
+/**
+ * @template TRead
+ * @template TWrite
+ * @template TReadReturn
+ * @template TWriteReturn
+ * @param {import('./types.js').Stream<TWrite, TRead, TWriteReturn, TReadReturn>} writer
+ * @param {import('./types.js').Stream<TRead, TWrite, TReadReturn, TWriteReturn>} reader
+ * @param {TWrite} primer
+ */
+export const pump = async (writer, reader, primer) => {
+  /** @param {Promise<IteratorResult<TRead, TReadReturn>>} promise */
+  const tick = promise =>
+    promise.then(
+      result => {
+        if (result.done) {
+          return writer.return(result.value);
+        } else {
+          // Behold: mutual recursion.
+          // eslint-disable-next-line no-use-before-define
+          return tock(writer.next(result.value));
+        }
+      },
+      (/** @type {Error} */ error) => {
+        return writer.throw(error);
+      },
+    );
+  /** @param {Promise<IteratorResult<TWrite, TWriteReturn>>} promise */
+  const tock = promise =>
+    promise.then(
+      result => {
+        if (result.done) {
+          return reader.return(result.value);
+        } else {
+          return tick(reader.next(result.value));
+        }
+      },
+      (/** @type {Error} */ error) => {
+        return reader.throw(error);
+      },
+    );
+  await tick(reader.next(primer));
+  return undefined;
+};
+harden(pump);
+
+/**
+ * @template TRead
+ * @template TWrite
+ * @template TReturn
+ * @param {AsyncGenerator<TRead, TReturn, TWrite>} generator
+ * @param {TWrite} primer
+ */
+export const prime = (generator, primer) => {
+  // We capture the first returned promise.
+  const first = generator.next(primer);
+  const primed = harden({
+    /** @param {TWrite} value */
+    async next(value) {
+      const result = await first;
+      if (result.done) {
+        return result;
+      }
+      return generator.next(value);
+    },
```

Comment:
The `await` here causes an extra turn *each time* even though it can only be relevant the first time.

True as well for `return` and `throw`, but not a big deal there because they're normally only called one anyway. Nevertheless, if you used local state to remember the fulfillment of `first`, you could avoid all these unnecessary turns.

Reply:
Got this and managed to not make the mistake of calling `next` on concurrent cache misses. Just once up front!